refactor(compiler): rework and expose APIs to be used in schematics

crisbeto opened the pull request:
Reworks some of the existing compiler APIs to make them easier to use in a schematic and exposes a few new ones to surface information we already had. High-level list of changes:
* `getPotentialImportsFor` now requires a class reference, instead of a `PotentialDirective | PotentialPipe`.
* New `getNgModuleMetadata` method has been added to the type checker.
* New `getPipeMetadata` method has been added to the type checker.
* New `getUsedDirectives` method has been added to the type checker.
* New `getUsedPipes` method has been added to the type checker.
* The `decorator` property was exposed on the `TypeCheckableDirectiveMeta`. The property was already present at runtime, but it wasn't specified on the interface.

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -732,7 +757,7 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
     if (exportingNgModules !== null) {
       for (const exporter of exportingNgModules) {
-        const emittedRef = this.emit(PotentialImportKind.Standalone, exporter, inContext);
+        const emittedRef = this.emit(PotentialImportKind.NgModule, exporter, inContext);
```
@dylhunn this change is unrelated to what I need for the schematics, but it seemed like an oversight from the previous PR, unless I'm misunderstanding it.

dylhunn replied:
Yes, this does seem like an error. Curious it didn't cause the previous PR to misbehave.

crisbeto commented on packages/compiler/src/render3/view/t2_binder.ts:
```diff
@@ -331,8 +331,6 @@ class DirectiveBinder<DirectiveT extends DirectiveMeta> implements Visitor {
 class TemplateBinder extends RecursiveAstVisitor implements Visitor {
   private visitNode: (node: Node) => void;
 
-  private pipesUsed: string[] = [];
```
Unrelated change. I just noticed that it's not being used.

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/api/checker.ts:
```diff
@@ -150,8 +153,8 @@ export interface TemplateTypeChecker {
-      toImport: PotentialDirective|PotentialPipe,
-      inComponent: ts.ClassDeclaration): ReadonlyArray<PotentialImport>;
+      toImport: Reference<ClassDeclaration>, inComponent: ts.ClassDeclaration,
+      asDirectImport?: boolean): ReadonlyArray<PotentialImport>;
```
I'm not a fan of the `asDirectImport` parameter, but I couldn't think of a better way to call it. I need it in the schematic, because I know that some directives that currently aren't standalone will become standalone as a part of the migration so I want the imports to be generated accordingly.

dylhunn replied:
I agree it's not the best API, but if you need it then you need it. Can you add a default option of `false` and give it a scarier name so it's not used elsewhere? Maybe something like `forceStandaloneImport`, and then add a `TODO` comment that it can be removed in a standalone world :)

dylhunn commented:
Thanks, reviewing this today.

dylhunn commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -598,6 +606,20 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
+    if (!isNamedClassDeclaration(module)) {
+      return null;
+    }
+    return this.metaReader.getNgModuleMetadata(new Reference(module));
```
Creating a new reference to perform the lookup works fine for locally declared ngModules, but may not work for ngModules coming from external dts file (i.e. anything not in the current compilation unit). This also applies to the other `get*Metadata` methods. I'm not sure if this matters to you -- for more details, look in the dts metadata reader.

crisbeto replied:
It doesn't really matter for my case. FWIW, I was following what we were doing in `getDirectiveMetadata`.

dylhunn reviewed (approved):
LGTM with notes above.

crisbeto committed:
refactor(compiler): rework and expose APIs to be used in schematics

Reworks some of the existing compiler APIs to make them easier to use in a schematic and exposes a few new ones to surface information we already had. High-level list of changes:
* `getPotentialImportsFor` now requires a class reference, instead of a `PotentialDirective | PotentialPipe`.
* New `getNgModuleMetadata` method has been added to the type checker.
* New `getPipeMetadata` method has been added to the type checker.
* New `getUsedDirectives` method has been added to the type checker.
* New `getUsedPipes` method has been added to the type checker.
* The `decorator` property was exposed on the `TypeCheckableDirectiveMeta`. The property was already present at runtime, but it wasn't specified on the interface.

crisbeto committed:
fixup! refactor(compiler): rework and expose APIs to be used in schematics

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -712,18 +734,22 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
-      toImport: PotentialDirective|PotentialPipe,
-      inContext: ts.ClassDeclaration): ReadonlyArray<PotentialImport> {
+      toImport: Reference<ClassDeclaration>, inContext: ts.ClassDeclaration,
+      forceStandaloneImport = false): ReadonlyArray<PotentialImport> {
```
```suggestion
      forceStandaloneImport): ReadonlyArray<PotentialImport> {
```
General style for the compiler is to avoid default values for parameters.

Instead of a boolean parameter, you might consider something like:

```typescript
export enum PotentialImportMode {
  Normal,
  ForceDirect,
}
```

as it carries more semantic information than a boolean flag.

crisbeto committed:
fixup! refactor(compiler): rework and expose APIs to be used in schematics

AndrewKushnir commented:
This PR was merged into the repository by commit 6beff5e8d708b575c38320fc8f997bd304a738f3.
